avalon/tools/widgets.py

```diff
@@ -0,0 +1,327 @@
+import logging
+
+from .lib import (
+    preserve_expanded_rows,
+    preserve_selection,
+    iter_model_rows
+)
```

Comment:
Thanks David. Is that based on the [e3 example in `CONTRIBUTING.md`](https://github.com/getavalon/core/blob/master/CONTRIBUTING.md#e3)? I interpreted it as *not* importing as a shorter name and not necessarily removing all functions imports.

But probably @mottosso can best describe his intentions. I'm fine with both, do note however that if this needs to be adapted that it accounts for more things, e.g. [see these lines](https://github.com/getavalon/core/blob/27449a5d809f1d325ed474d50d2388c93723a96a/avalon/tools/widgets.py#L8).

I did now see there's an import there that states:
```python
from ..vendor import qtawesome as awesome
```

Which I believe should be:
```python
from ..vendor import qtawesome
```

Reply:
That irritates me for a long time I think it should be `from ..vendor import qtawesome`

Reply:
> `from ..vendor import qtawesome as awesome`

This has now been resolved with [this commit](https://github.com/getavalon/core/pull/440/commits/9bc098edc701f9856d1b9ba0e016ddf22577fcb2).

Reply:
> Is that based on the e3 example in CONTRIBUTING.md?

I was actually referencing from [A8 in `CONTRIBUTING.md`](https://github.com/getavalon/core/blob/master/CONTRIBUTING.md#code-quality---architecture), which aim to preserve the namespace :P

> Especially now that there's a lib at multiple levels (avalon-wide + tools-wide + per tool) seeing lib. wouldn't tell much.

Yeah.. maybe this will work ? Say we are in bad luck, used all three levels' `lib.py` in one module:
```python
from ... import lib as avalon_lib
from .. import lib as tool_lib
from . import lib
```

Reply:
This one's a hard problem.

On the one hand, word and line length obscure code. One the other, namespaces give the reader a hint at what a function or class is coming from.

For some modules, it's obvious.

```py
qtawesome.some_function()  # I know you're third-party
```

For others, not so much.

```py
model = SomeModel()  # Third-party, local module, separate root-level module?
```

Let's try this as a general guideline; preserve the leaf-level module namespace.

- `from .vendor import qtawesome` is Good
- `from .vendor.qtawesome import icon` is Bad

> maybe this will work ?

This is another hard problem; on the one hand, we should use `lib` at various levels, as they are context sensitive. A `lib` under `tools` is tools libraries, whereas a `lib` under `maya/` is Maya libraries. And on the other they complicate the above general guideline.

I think this is as elegant of a solution as we'll get, if we are to have at least some guideline for how to deal with it. Preserve the *minimum* number of namespaces; in this case, we need both `tool` and `lib`, if `avalon_lib` is also present.

Reply:
Thanks @mottosso - just to be complete, if I understand you correctly the above example to import as `tool_lib`, etc. when required is considered the correctly approved change. Yes? 👍